File: esports-backend/src/main/java/com/esports/service/S3Service.java
package com.esports.service;

import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;
import software.amazon.awssdk.core.sync.RequestBody;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.GetObjectRequest;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;
import software.amazon.awssdk.services.s3.presigner.S3Presigner;
import software.amazon.awssdk.services.s3.presigner.model.GetObjectPresignRequest;

import java.io.IOException;
import java.time.Duration;
import java.util.Map;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class S3Service {
    private final S3Client s3Client;
    private final S3Presigner s3Presigner;

    @Value("${aws.s3.bucket}")
    private String bucket;

    public Map<String, String> upload(MultipartFile file, String folder)
            throws IOException {

        if (file.isEmpty()) {
            throw new IllegalArgumentException("File is empty");
        }

        // Allow larger files for project uploads (3GB limit)
        long maxSize = folder.startsWith("projects/") ? 3L * 1024 * 1024 * 1024 : 5L * 1024 * 1024;
        if (file.getSize() > maxSize) {
            String sizeLimit = folder.startsWith("projects/") ? "3GB" : "5MB";
            throw new IllegalArgumentException("Max file size is " + sizeLimit);
        }

        String safeFileName = UUID.randomUUID().toString() + "." + getFileExtension(file.getOriginalFilename());
        String key = folder + "/" + safeFileName;

        PutObjectRequest request = PutObjectRequest.builder()
                .bucket(bucket)
                .key(key)
                .contentType(file.getContentType())
                .build();

        s3Client.putObject(
                request,
                RequestBody.fromInputStream(
                        file.getInputStream(),
                        file.getSize()
                )
        );

        return Map.of(
                "key", key,
                "url", key  // Store just the key, generate presigned URL when needed
        );
    }

    public String generateDownloadUrl(String s3Key) {
        GetObjectRequest getObjectRequest = GetObjectRequest.builder()
                .bucket(bucket)
                .key(s3Key)
                .build();

        GetObjectPresignRequest presignRequest =
                GetObjectPresignRequest.builder()
                        .signatureDuration(Duration.ofMinutes(15))
                        .getObjectRequest(getObjectRequest)
                        .build();

        return s3Presigner
                .presignGetObject(presignRequest)
                .url()
                .toString();
    }

    private String getFileExtension(String filename) {
        if (filename == null || filename.lastIndexOf('.') == -1) {
            return "jpg";
        }
        return filename.substring(filename.lastIndexOf('.') + 1);
    }
}

File: Upkart1/Services/S3Service.cs
using Amazon;
using Amazon.S3;
using Amazon.S3.Model;

namespace Upkart1.Services
{
    public class S3Service : IS3Service
    {
        private readonly IAmazonS3 _s3Client;
        private readonly string _bucket;

        public S3Service(IConfiguration config)
        {
            _bucket = config["AWS:S3:Bucket"];
            _s3Client = new AmazonS3Client(
                config["AWS:AccessKeyId"],
                config["AWS:SecretAccessKey"],
                RegionEndpoint.GetBySystemName(config["AWS:Region"])
            );
        }

        public async Task<Dictionary<string, string>> Upload(IFormFile file, string folder)
        {
            if (file.Length == 0)
                throw new ArgumentException("File is empty");

            // Allow larger files for project uploads (3GB limit)
            long maxSize = folder.StartsWith("projects/") ? 3L * 1024 * 1024 * 1024 : 5L * 1024 * 1024;
            if (file.Length > maxSize)
            {
                string sizeLimit = folder.StartsWith("projects/") ? "3GB" : "5MB";
                throw new ArgumentException($"Max file size is {sizeLimit}");
            }

            var safeFileName = Guid.NewGuid().ToString() + "." + GetFileExtension(file.FileName);
            var key = $"{folder}/{safeFileName}";

            using var stream = file.OpenReadStream();

            var request = new PutObjectRequest
            {
                BucketName = _bucket,
                Key = key,
                InputStream = stream,
                ContentType = file.ContentType
            };

            await _s3Client.PutObjectAsync(request);

            return new Dictionary<string, string>
            {
                ["key"] = key,
                ["url"] = key  // Store just the key, generate presigned URL when needed
            };
        }

        public string GenerateDownloadUrl(string s3Key)
        {
            var request = new GetPreSignedUrlRequest
            {
                BucketName = _bucket,
                Key = s3Key,
                Expires = DateTime.UtcNow.AddMinutes(15)
            };

            return _s3Client.GetPreSignedURL(request);
        }

        public string GetDownloadUrl(string s3Key)
        {
            return GenerateDownloadUrl(s3Key);
        }

        private string GetFileExtension(string filename)
        {
            if (filename == null || filename.LastIndexOf('.') == -1)
                return "jpg";
            return filename.Substring(filename.LastIndexOf('.') + 1);
        }
    }
}
